compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13314,113 +13577,92 @@ OMR::Z::TreeEvaluator::primitiveArraycopyEvaluator(TR::Node* node, TR::CodeGener
+   TR::Instruction *cursor = NULL;
+   TR::RegisterDependencyConditions *deps = NULL;
+   TR_Debug *debug = cg->getDebug();
+   // If it is variable length load then Internal Control Flow starts here where on runtime we check if length of bytes to copy is greater than zero.
+   if (!isConstantByteLen)
       {
-      // The OOL path needs the byte length for the guarded load, compress, and store loop.
-      // However, the main line fast path also needs the byte length. As such, the byteLenNode
-      // must be evaluated outside of the OOL ICF region.
       byteLenReg = cg->gprClobberEvaluate(byteLenNode);
-      alreadyEvaluatedByteLenNode = true;
-
-      mergeLabel = generateLabelSymbol(cg);
-      TR::LabelSymbol* slowPathLabel = generateLabelSymbol(cg);
-
-      TR_ASSERT_FATAL(J9_PRIVATE_FLAGS_CONCURRENT_SCAVENGER_ACTIVE == 0x20000,
-                 "GSCS: The OOL sequence branch is dependant on the flag being 0x20000");
-      TR_ASSERT_FATAL(TR::Compiler->target.is64Bit(),
-                 "GSCS: Guarded Load OOL Path only defined in 64 bit mode");
-      TR::Register *vmReg = cg->getMethodMetaDataRealRegister();
-      TR::MemoryReference *privFlagMR = generateS390MemoryReference(vmReg,
-            TR::Compiler->vm.thisThreadGetConcurrentScavengeActiveByteAddressOffset(cg->comp()), cg);
-      generateSIInstruction(cg, TR::InstOpCode::TM, node, privFlagMR, 0x00000002);
-      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_CC3, node, slowPathLabel);
-
-      // Generate OOL Slow Path
-      TR_S390OutOfLineCodeSection* outOfLineCodeSection = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(slowPathLabel, mergeLabel, cg);
-      cg->getS390OutOfLineCodeSectionList().push_front(outOfLineCodeSection);
-      outOfLineCodeSection->swapInstructionListsWithCompilation();
-
-      generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, slowPathLabel);
-
-      TR::Register* strideRegister = cg->allocateRegister();
-
-      MemCpyVarLenTypedMacroOp vtOpSlow(node, byteDstNode, byteSrcNode, cg, node->getArrayCopyElementType(), byteLenReg, byteLenNode, true, node->isForwardArrayCopy());
-      vtOpSlow.generate(byteDstReg, byteSrcReg, strideRegister);
-      vtOpSlow.cleanUpReg();
-
-      cg->stopUsingRegister(strideRegister);
-
-      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, mergeLabel);
-
-      // Switch back from OOL Slow Path
-      outOfLineCodeSection->swapInstructionListsWithCompilation();
+      generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, startICFLabel);
+      startICFLabel->setStartInternalControlFlow();
+      cursor = generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::getCmpOpCode(), node, byteLenReg, static_cast<int8_t>(0x0), TR::InstOpCode::COND_BNH, mergeLabel, false);
+      iComment("byteLen <= 0 goto mergeLabel");
       }
-#endif
 
    if (node->isForwardArrayCopy())
```

Comment:
Don't we also have `isBackwardArrayCopy()`? Are we able to make use of this to eliminate the runtime check?